content/advent-2018/go-arrow.md

```diff
@@ -0,0 +1,531 @@
+
+type array struct {
+	refCount        int64
+	data            *Data
+	nullBitmapBytes []byte
+}
+```
+
+Creating a new array value holding `int32`s is done with `array.NewInt32Data(data)` where `data` holds a carefully crafted `array.Data` value with the needed memory buffers and sub-buffers (for the `null` bitmap.)
+As this can be a bit unwiedly, a helper type is provided to create Arrow arrays from scratch:
+
+```go
+mem := memory.NewGoAllocator()
+bld := array.NewInt32Builder(mem)
+defer bld.Release()
+
+// create an array with 4 values, no null
+bld.AppendValues([]int32{1, 2, 3, 4}, nil)
+
+arr1 := bld.NewInt32Array() // materialize the array
+defer arr1.Release()        // make sure we release memory, eventually.
+
+// arr1 = [1 2 3 4]
+fmt.Printf("arr1 = %v\n", arr1)
+
+// create an array with 5 values, 1 null
+bld.AppendValues(
+	[]int32{1, 2, 3, 4, 5},
+	[]bool{true, true, true, false, true},
+)
+
+arr2 := bld.NewInt32Array()
+defer arr2.Release()
+
+// arr2 = [1 2 3 (null) 5]
+fmt.Printf("arr2 = %v\n", arr2)
+```
+
+Arrow arrays can also be sub-sliced:
+
+```go
+sli := array.NewSlice(arr2, 1, 5).(*array.Int32)
+defer sli.Release()
+
+// slice = [2 3 (null) 5]
+fmt.Printf("slice = %v\n", sli)
+```
+
+Similar code can be written for arrays of `float64`, `float32`, unsigned integers, etc...
+But not everything can be mapped or simply expressed in terms of these basic primitives.
+What if we need a more detailed data type?
+That is expressed with `List`s or `Struct`s.
+Let us imagine we need to represent an array of "Person" type like:
+```go
+type Person struct {
+	Name string
+	Age  int32  // presumably, int8 should suffice.
+}
+```
+
+Structured types in Arrow can be seen as entries in a database, with a specified schema.
+While many databases are row oriented, data is column-oriented in Arrow:
+
+![simd](/postimages/advent-2018/go-arrow/arrow-simd.png)
+
+The reasoning behind organizing data along columns instead of rows is that:
+
+- compression should work better for data of the same type, (presumably, data should be relatively similar)
+- many workloads only care about a few columns to perform their analyses, the other columns can be left at rest on disk,
+- one can leverage [vectorized instructions](https://en.wikipedia.org/wiki/SIMD) to carry operations.
+
+This is the well-known _structure of arrays_ (`SoA`) _vs_ _arrays of structure_ (`AoS`) [debate](https://en.wikipedia.org/wiki/AOS_and_SOA).
+
+With Arrow, the `Person` type defined previously could be implemented as the following structure of arrays:
+
+```go
+mem := memory.NewGoAllocator()
+dtype := arrow.StructOf([]arrow.Field{
+	{Name: "Name", Type: arrow.ListOf(arrow.PrimitiveTypes.Uint8)},
+	{Name: "Age",  Type: arrow.PrimitiveTypes.Int32},
+}...)
+
+bld := array.NewStructBuilder(mem, dtype)
+defer bld.Release()
+```
+
+A complete example is provided [here](https://godoc.org/github.com/apache/arrow/go/arrow#example-package--StructArray).
+
+## Arrow tables
```

Comment:
I guess I missed this in code reviews. Does the Go library have a "record batch" notion? In C++/Python we distinguish between a record batch (all columns are contiguous single arrays) and tables (where each column can be chunked in one or more arrays). It seems here that the Go Table is the latter, which is consistent. 

From a binary protocol / data sharing perspective, there are only record batches, and tables are a purely semantic / programmatic construct

Reply:
yes. I followed the C++ implementation and defined a `Record` interface:
- https://godoc.org/github.com/apache/arrow/go/arrow/array#Record

for this blog post, I felt it was easier to just introduce/focus-on tables as that's a term people already know, even if there were some subtle nuances.